Add hook findDuplicates

mattwire opened the pull request:
Overview
----------------------------------------
I need an extension to be able to override the duplicate checking when registering for an event.  Specifically, based on the context (ie. type of event, custom fields linked to event), I need to handle the duplicate checking in a different way.  Currently there is no way to change this handling without overriding core files.

This is a proposal to add a hook that allows the duplicate checking to be intercepted.  Context (currently just for event) is additionally passed into the hook and the hook returns a parameter "handled" which allows us to skip or run the core duplicate checking.

Before
----------------------------------------
Not possible to override behaviour of duplicate checking (you can select which rule should be used for an event but this does not allow for more custom conditions to be considered, such as in what "area" the event is being run.

After
----------------------------------------
Hook provides ability for extensions to intercept and/or override core duplicate checking when registering/contributing.

Technical Details
----------------------------------------
_If the PR introduces noteworthy technical changes, please describe them here. Provide code snippets if necessary_

Comments
----------------------------------------
@seamuslee001 @eileenmcnaughton @monishdeb Feedback appreciated!

eileenmcnaughton commented:
In general there are hooks in the duplicates code that are in bad places,  take inappropriate parameters & make it really hard to clean up code. This looks like a place that has already been somewhat cleaned up but let's be really careful about what we pass through because once people are relying on things in hooks we are a bit stuck

mattwire commented:
> DedupeFinderTest.php:210, HookComment, Priority: Normal
> --
> Format should be "* Implements hook_foo().", "* Implements hook_foo_BAR_ID_bar() for xyz_bar().",, "* Implements hook_foo_BAR_ID_bar() for xyz-bar.html.twig.", or "* Implements hook_foo_BAR_ID_bar() for xyz-bar.tpl.php.".


@seamuslee001 The checkstyle warning is doing my head in!  I can't seem to get the comment in a format that jenkins is happy with - any thoughts?

seamuslee001 commented on tests/phpunit/CRM/Dedupe/DedupeFinderTest.php:
```diff
@@ -204,6 +206,43 @@ public function testDupesByParams() {
   }
 
+  /**
+   * Implements hook_civicrm_findDuplicates().
```
try adding a blank line with just a * in after this

seamuslee001 commented on tests/phpunit/CRM/Dedupe/DedupeFinderTest.php:
```diff
@@ -204,6 +206,43 @@ public function testDupesByParams() {
+   * Implements hook_civicrm_findDuplicates().
+   * Locks in expected params
+   *
+   * @param array $dedupeParams
```
it'll probably complain your duplicating documentation here

colemanw commented:
@eileenmcnaughton I'm not sure what to do with this PR based on your comment. In your opinion should it be closed or is it workable?

mattwire commented:
@colemanw When @eileenmcnaughton first commented I was passing in the whole event detail array - I've now reduced that to just pass in "event_id" in the context and included that parameter in the test.  Hopefully that addresses the concerns.

There is no other way to intercept this part of the code, which is used during registration/contribution for contact matching or creating a new contact.

monishdeb commented:
Also @mattwire can you submit a doc PR in order to add new documentation for this new hook in [here](https://github.com/civicrm/civicrm-dev-docs/tree/master/docs/hooks)?  (would appreciate if you can cite an example in that doc)

mattwire commented:
@monishdeb I've added a docs PR.  @eileenmcnaughton Do you think this is ok now?

eileenmcnaughton commented:
OK - I think this is - it still worries me that people might add weird & wonderful things to contextParams but OK.... (in other places I've done some cruel things on having a array of testedParams & filtering on them but ...)

eileenmcnaughton commented:
@mattwire where is the docs PR?

mattwire commented:
I thought I'd linked it - docs PR here: https://github.com/civicrm/civicrm-dev-docs/pull/584
